### ISSUE TITLE
Processes are not shut down when windows using a WebView are closed

### ISSUE THREAD

**Issue by cremor** [MicrosoftEdge/WebView2Feedback#1424]
**Description**
In a WPF application that opens windows that contain a WebView2 control each opened window starts a new WebView2 process, but closing windows does not close the additional process. This means that after a while a huge number of WebView2 processes could be running and unnecessarily use memory.

**Version**
SDK: 1.0.864.35 (same behavior with 1.0.902-prerelease)
Runtime: Evergreen 91.0.864.53
Framework: WPF
OS: Windows 10 x64 21H1 (but the WPF app is x86)

**Repro Steps**
1. Create a WPF application that can open additional windows.
2. Open a window that contains a WebView control just like described in ["Step 3 - Create a single WebView" of the WPF getting started guide](https://docs.microsoft.com/en-us/microsoft-edge/webview2/get-started/wpf#step-3---create-a-single-webview).
3. Close the window again.
3. Repeat steps 2 and 3 a few times.

Expected behavior:
The number of running WebView2 processes should stay the same, regardless of how many windows are opened and closed. (Shared processes should could continue running, but window-specific processes should shut down.)

Actual behavior:
Opening new windows starts new WebView2 processes, but closing windows doesn't shut them down. This is a memory leak.

**Screenshots**
After a while it looks like this, even if all windows containing WebView2 controls are already closed:
![grafik](https://user-images.githubusercontent.com/116314/122566964-6ffce700-d048-11eb-9aff-d4e9389c00cf.png)

**Additional context**
I've tried adding a call to `WebView2.Dispose()` in the `Window.OnClosed()` method. This prevents the memory leak, but it has a big drawback:
When the last WebView2 control is disposed, all of the WebView2 processes are shut down. So not only the WebView2 is disposed, but also the CoreWebView2 (this is even [documented](https://docs.microsoft.com/en-us/dotnet/api/microsoft.web.webview2.wpf.webview2.dispose?view=webview2-dotnet-1.0.864.35)). But I'd like to keep the CoreWebView2 alive because creating the CoreWebView2 is an expensive operation as documented [here](https://docs.microsoft.com/en-us/dotnet/api/microsoft.web.webview2.wpf.webview2?view=webview2-dotnet-1.0.864.35).

In other words:
WPF windows that contain WebView2 controls should behave like tabs in the Edge application. Closing windows should shut down the processes associated with that window, just like closing tabs in Edge shuts down those processes. But the shared processes should stay running until the main process is shut down.

**Comment by champnic** [comment]
Hey @cremor - It is currently By Design that the WebView2 must be disposed before it cleans up the associated renderer process. It also matches browser behavior that when all WebView2's are closed, the shared browser processes are shut down - when you close all tabs in a browser window, the window itself is also closed and the browser processes exit.

If you'd like to change this behavior you can keep around a hidden WebView2 (navigated to about:blank, for example). You can call TrySuspend on that WebView2 to limit it's resource consumption while it's not in use. Thanks!

**Comment by cremor** [comment]
> If you'd like to change this behavior you can keep around a hidden WebView2 (navigated to about:blank, for example). You can call TrySuspend on that WebView2 to limit it's resource consumption while it's not in use. Thanks!

@champnic Is this possible without actually adding the hidden WebView2 to a XAML file? I've tried it in code and it seems like it never is actually initialized (it never starts the renderer process and doesn't prevent disposing of the CoreWebView2).
This is also the case if I explicitly call `EnsureCoreWebView2Async()`. The task returned from that method is never completed.

**Comment by champnic** [comment]
The WebView2 would need a valid HWND parent from the host app, which in WPF is done using HwndHost so I believe it would have to be somewhere in the visual tree for this to work. This doesn't necessarily mean the XAML file (you could do it in code) and you would keep it hidden.

**Comment by cremor** [comment]
@champnic  Thanks. I've managed to do this now but it wasn't easy. In case anyone else needs that functionality, the following points are important:

* The background WebView2 needs to be contained in a Window which is shown - although it can be invisible. Just calling `WindowInteropHelper.EnsureHandle()` is **not** enough to initialize the WebView2.
* The background WebView2 needs to use the same UserDataFolder as the "real" WebView2, otherweise a whole separate background process tree is started.
* To make sure point 2 actually works you have to set the `WebView2.CreationProperties` **before** you set `WebView2.Source`.

And here is my solution. If someone knows how to improve (or shorten) this, please tell.

```cs
// Called in CoreWebView2InitializationCompleted of the "real" WebView2
private static void EnsureBackgroundWebViewIsActive() {
   if (_backgroundWebView != null) {
      return;
   }

   _backgroundWebView = new WebView2();
   SetWebViewCreationProperties(_backgroundWebView);
   _backgroundWebView.Source = new Uri("about:blank");

   var window = new Window {
      Content = _backgroundWebView,
      Width = 0,
      Height = 0,
      WindowStyle = WindowStyle.None,
      ShowInTaskbar = false,
      ShowActivated = false
   };
   window.Show();
   window.Close();
}
```

This would actually need even more code. E.g. `ProcessFailed` should be handled to restart the background WebView2. But for now that's how I do it.

**Comment by AndyT-MS** [comment]
@cremor Another idea that you could try if you want to keep some of the non-render processes around for later use is to manage a `CoreWebView2Environment` instance yourself so that you can control its lifetime.  I'm not an expert in this area and haven't actually tried it myself, so *don't take this as any kind of authoritative solution*, but I think that at least some of those processes are tied to the lifetime of the environment.

To implement this idea, create your own environment via `CoreWebView2Environment.CreateAsync(...)` before you initialize your first `WebView2` control, perhaps when your app first starts up.  Then whenever you're initializing a `WebView2` control pass your environment instance to its `EnsureCoreWebView2Async(...)` method.  Keep the environment instance around even after your controls are gone and/or `Dispose`d and keep using it when you initialize new ones.  Maybe just keep it around until your app shuts down.

Note that for this to work you'll need to make sure that you call `EnsureCoreWebView2Async` on any new `WebView2` control *before* you set its `Source` property, which means that you can't set its `Source` property in XAML.  I suggest doing that during its `CoreWebView2InitializationCompleted` event instead.

**Comment by dacurrey** [comment]
To reuse a Webview2 control from a previously closed window.

1. Ensure that the Webview2 control is initialized from the Main Window
2. Just add a call to InitializeWeb21 when you reopen a closed window

    Private Sub OpenAnotherWindow()
        If PreviouslyLoaded = True Then InitializeWebView21()
    End Sub


    Private Async Sub InitializeWebView21()
        Await Form2.WebView21.EnsureCoreWebView2Async()
    End Sub
